src/xdp-app-info-registry.c

```diff
@@ -0,0 +1,174 @@
+
+  GHashTable *app_infos; /* unique dbus name -> app info */
+  GMutex app_infos_mutex;
+};
+
+G_DEFINE_FINAL_TYPE (XdpAppInfoRegistry,
+                     xdp_app_info_registry,
+                     G_TYPE_OBJECT)
+
+static void
+xdp_app_info_registry_dispose (GObject *object)
+{
+  XdpAppInfoRegistry *registry = XDP_APP_INFO_REGISTRY (object);
+
+  if (registry->app_infos)
+    {
+      g_mutex_clear (&registry->app_infos_mutex);
+      g_clear_pointer (&registry->app_infos, g_hash_table_unref);
+    }
+
+  G_OBJECT_CLASS (xdp_app_info_registry_parent_class)->dispose (object);
+}
+
+static void
+xdp_app_info_registry_class_init (XdpAppInfoRegistryClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->dispose = xdp_app_info_registry_dispose;
+}
+
+static void
+xdp_app_info_registry_init (XdpAppInfoRegistry *registry)
+{
+}
+
+XdpAppInfoRegistry *
+xdp_app_info_registry_new (void)
+{
+  XdpAppInfoRegistry *registry = g_object_new (XDP_TYPE_APP_INFO_REGISTRY, NULL);
+
+  registry->app_infos = g_hash_table_new_full (g_str_hash, g_str_equal,
+                                               g_free,
+                                               g_object_unref);
+  g_mutex_init (&registry->app_infos_mutex);
+
+  return registry;
+}
+
+XdpAppInfo *
+xdp_app_info_registry_lookup_sender (XdpAppInfoRegistry *registry,
+                                     const char         *sender)
+{
+  XdpAppInfo *app_info = NULL;
+
+  g_mutex_lock (&registry->app_infos_mutex);
+  app_info = g_hash_table_lookup (registry->app_infos, sender);
+  if (app_info)
+    g_object_ref (app_info);
+  g_mutex_unlock (&registry->app_infos_mutex);
+
+  return app_info;
+}
+
+gboolean
+xdp_app_info_registry_has_sender (XdpAppInfoRegistry *registry,
+                                  const char         *sender)
+{
+  gboolean has_app_info = FALSE;
+
+  g_mutex_lock (&registry->app_infos_mutex);
+  has_app_info = !!g_hash_table_lookup (registry->app_infos, sender);
+  g_mutex_unlock (&registry->app_infos_mutex);
+
+  return has_app_info;
+}
+
+void
+xdp_app_info_registry_insert (XdpAppInfoRegistry *registry,
+                              XdpAppInfo         *app_info)
+{
+  const char *sender = xdp_app_info_get_sender (app_info);
+
+  g_debug ("Adding XdpAppInfo: %s app '%s' for %s",
+           xdp_app_info_get_engine_display_name (app_info),
+           xdp_app_info_get_id (app_info),
+           sender);
+
+  g_mutex_lock (&registry->app_infos_mutex);
+  g_hash_table_insert (registry->app_infos,
+                       g_strdup (sender),
+                       g_object_ref (app_info));
+  g_mutex_unlock (&registry->app_infos_mutex);
+}
+
+void
+xdp_app_info_registry_delete (XdpAppInfoRegistry *registry,
+                              const char         *sender)
+{
+  XdpAppInfo *app_info = NULL;
+
+  g_mutex_lock (&registry->app_infos_mutex);
+  app_info = g_hash_table_lookup (registry->app_infos, sender);
+  if (app_info)
```

Comment:
Nit: you can save the lookup by just putting `if (g_hash_table_remove (..., sender))` directly in the condition